src/Common/src/Internal/NativeFormat/NativeFormatReader.Primitives.cs

```diff
@@ -25,14 +25,27 @@ public static byte ReadUInt8(ref byte* stream)
 
         public static ushort ReadUInt16(ref byte* stream)
         {
+#if WASM && DEBUG // WASM debug may output to asm.js which only supports aligned access
```

Comment:
I think we should fix this by doing [Unsafe.ReadUnaligned](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.readunaligned?view=dotnet-plat-ext-2.1). But we'll need to think through how to do this in a clean way. This file is included from a bunch of places. Some of them have access to the `Internal.Runtime.CompilerServices.Unsafe` class.

Maybe it wouldn't be too terrible to just put a `using Internal.Runtime.CompilerServices.Unsafe` on top of the file and use the Unsafe from the internal namespace. CSPROJ files that don't include it yet but should can just source include it.

We currently have two copies of Unsafe class. One is in System.Private.CoreLib, the other is in Runtime.Base. I would move the one from Runtime.Base to /src/Common, and add the missing implementation of ReadUnaligned (steal it from the copy in CoreLib). We can source include that one from places that need it.

I expect it might be 2-3 iterations to get it right.

Reply:
I'm happy to go ahead with this, please don't take this as I've not got the enthusiasm, but I have this nagging doubt even though I recently suggested we try to get these changes in.  Blazor dropped support for asm.js, I think because mono dropped support for asm.js perhaps because of the requirement to align access.   These changes may be useful for other targets like Qualcomm, I'm not an expert in which CPUs would benefit from, or require aligning word access, so I just wanted to canvas an opinion on how valuable this is and if the only advantage is to debug (and support IE which only has a couple of years of support), then would it be better to focus on debugging wasm rather than asm.js?

Reply:
We do like out core to be portable. Ideally, this would just use `BinaryPrimitives.ReadInt16LittleEndian` that deals with all the concerns here.

`BinaryPrimitives.ReadInt16LittleEndian` exists in netcoreapp only. It does not exist in `netstandard1.3` that a lot of the projects still builds against. I think updating everything to netcoreapp would be larger undertaking that you would like to take.

To make progress on this, I would change this to:

```
#if netstandard
    ushort result = *(ushort*)(stream); // Assumes little endian and unaligned access
#else
    ushort result = BinaryPrimitives.ReadInt16LittleEndian(new ReadOnlySpan<byte>(stream, 4));
#endif
```

and make sure that `netstandard` is defined when we are building against netstandard.

> would it be better to focus on debugging wasm rather than asm.js?

Agree with this too.